---
title: "Simulating phenotypes and evaluating GWAS methods with naturalgwas"
author: "Olivier François"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

#### Summary

Association studies of polygenic traits with genomic loci are notoriously difficult when those studies are conducted at large geographical scales. The difficulty arises as genotype frequencies often vary in geographic space and across distinct environments in a complex fashion. Those large-scale variations are known to yield false positives in association testing approaches. The R package **naturalgwas** allows researchers to simulate trait association at weakly linked loci, modelling gene by environment interaction where environment is derived from a climate database.  This package includes an oracle method correcting false associations by using the true set of confounding variables. The simulated data and the oracle method can be used 1) to compare the ability of association methods to correctly remove confounding factors for a particular data set, 2) to evaluate power to detect causal variants, and 3) to assess the influence of various parameters including the number of causal variants, effect sizes and gene by environment interaction. 

----

#### Introduction

This vignette presents a short tutorial on how to simulate phenotypes for "natural" organisms, and evaluate genome-wide association studies (GWAS) when field sampling has been performed at a large geographical scale. We illustrate the simulation approach with some data simulated for the plant species *Arabidopsis thaliana* (Atwell et al. 2010). To start with the R package **naturalgwas**, load the R functions in memory space as follows.


```{r}
#install.packages("cate")
#devtools::install_github("bcm-uga/lfmm")
library(naturalgwas)
```



##### Data files

Running the main simulation function **simu_pheno**, and running the **oracle** method requires three files as input to the program: 1) a file encoding individual genotypes, 2) a file with individual geographic coordinates, and 3) a genetic map.  The simulation could also work without geographic data by specifying a \code{NULL} environment. For example, consider SNP data from chromosome 1 of European accession/ecotypes of the plant species *A. thaliana*.


```{r data} 
data(A.thaliana)
genotype <- A.thaliana$genotype
chrpos <- A.thaliana$chrpos 
```

For SNPs, the **genotype** matrix encodes each individual genotype in a row. Each locus corresponds to a specific column. Genotypes are encoded as 0,1,2 for diploids, and 0,1 for haploids. Those numbers represent the number of reference or derived allele at each particular locus. *A. thaliana* is a diploid species with very high levels of inbreeding. In our example, 170 genotypes were encoded as haploids considering 26,943 loci from chromosome 1 (Atwell et al. 2010). For our example, the genetic map is artificial, meaning that the SNP positions don't match with the plant genome. Let's print some genotypic values for the first three individuals at the first ten loci. We have a matrix of size 3x10 filled with 0 and 1 values.  


```{r}
dim(genotype)
genotype[1:3,1:10]
```

The **coordinate** variable is a two-column matrix that contains longitude and latitude for each individual in the sample. Longitude (°E) and latitude (°N) must be encoded in the decimal format. Note that headers should be ignored when loading data into the R program. Let us show the location of the sampling sites in Europe (each points corresponds to an individual in the sample).


```{r}
library(maps)
coordinates <- A.thaliana$coord
plot(coordinates, 
     pch = 19, cex = .7, col = "green4",
     xlab = "Longitude (°E)", 
     ylab = "Latitude (°N)")
map(add = T, interior = FALSE)
```



##### Simulating artificial phenotypes from real genotypes

For simulating artificial phenotypes from real genotypes and accounting for genotype by environment interactions, we need 1) a proxy variable for the enviromnent, 2) a reference set of weakly linked causal loci for which we wish to simulate trait association, 3) and a set of confounder factors typically arising from complex population structure and isolation-by-distance.  

```{r init, dependson=c("data"), results="hide"} 
  env <- get_climate(coordinates)
  ref.set <- create_refset(chrpos, window = 201)
  confounder <- create_factor(genotype, K = 10)
```  
  
The above commands create an artificial environmental variable by combining 19 bioclimatic variables extracted from the 'worldclim' database, and they take a few minutes to download. The SNP reference set consists of picking one representant SNP in each window of size **window**. In the third line, 10 confounders are computed from the data. The confounders correspond to the first ten principal components of the genotype matrix using PCA from LEA (Frichot and Francois 2015). The PCA screeplot for the SNP data is displayed.

In addition, the **create_factor** function estimates the amount of residual variation in the data and an order of magnitude for effect sizes.

```{r, dependson=c("init")} 
  confounder$sigma 
  confounder$base 
```  

Simulation of phenotypic traits requires the following arguments: 1) a SNP genotype matrix, 2) a set of confounders created with **create_factor**, an environmental variable, a reference set of weakly linked SNPs created with **create_refset**, a number of causal SNPs to be drawn from the reference set, and values for effect size and GxE interaction parameters. 

```{r sim, dependson=c("init"), results="hide"}   
  n.causal <- 6
  eff.size <- 5.0

  sim <- simu_pheno(A.thaliana$genotype, 
                    confounder, 
                    env, 
                    ref.set, ncausal = n.causal, 
                    effect.size = eff.size, gxe = .5)
  
  ss <- sum(apply(genotype[,sim$causal.set], 2, var))
  ss*(eff.size)^2*confounder$base^2/(ss*(eff.size)^2*confounder$base^2 + confounder$sigma^2 + confounder$base^2)
  (ss*(eff.size)^2*confounder$base^2 + confounder$sigma^2)/(ss*(eff.size)^2*confounder$base^2 + confounder$sigma^2 + confounder$base^2)
```

In this simulation, a polygenic trait is associated with 6 causal variants, sampled from the SNP reference set. Each causal variant is associated with an effect size of 5 units. Gene by environment interactions are modelled by using the *env* variable created from 'worldclim'. 


```{r, dependson=c("sim")}   
  hist(sim$phenotype, main = "Trait value")
```

To check whether the simulated trait displays any geographic variation, the trait variable can interpolated on a geographic map as follows. 

```{r, dependson=c("sim")}   
  library(fields)
  fit = fields::Krig(coordinates, sim$phenotype, theta = 10, m = 2)
  surface(fit, extrap = TRUE, xlab = "Longitude", ylab = "Latitude", levels = c(0))
  map(add = TRUE, interior = F)
```



##### Running the oracle method

Our next step is to perform a GWAS for the simulated phenotype, and to evaluate the relative power of methods to detect causal variants. To this objective, we use the 'linear regression' method as an oracle algorithm. The method is very close to the simulation model, and it becomes an oracle method when the true confounders are provided in arguments.


```{r oracle, dependson=c("sim")}   
pv.oracle <- oracle(sim$phenotype, 
                    genotype, 
                    confounder, K = 10)$pv
```

Now visualize the Manhattan plot for the oracle method. The vertical bars correspond to the causal variants in the simulation.

```{r, dependson=c("cate")}   
   plot(-log10(pv.oracle), cex = .4, col = "grey", main = "Manhattan plot (Oracle)")
   points(sim$causal, -log10(pv.oracle)[sim$causal], type = "h", lty = 1, col = "blue")
   abline(-log10(0.05/length(pv.oracle)), 0, lty = 2)
```

##### Mixed model and heritability

```{r}
library(rrBLUP)
G <- genotype
rownames(G) <- 1:170
data <- data.frame(y = sim$phenotype, 
                   gid = 1:170)

#predict breeding values and compute h2
kb <- kin.blup(data = data, 
                geno="gid", 
                pheno = "y", 
                K = A.mat(G))

cat("Heritability = ", kb$Vg/(kb$Vg + kb$Ve), "\n")
```

```{r}
G <- data.frame(marker = 1:26943,
                A.thaliana$chrpos,
                t(genotype),
                check.names = FALSE)
pheno <- data.frame(line=1:170, y = sim$phenotype)

mlm <- GWAS(pheno = pheno, 
            geno = G, 
            K=NULL,
            plot=FALSE)
```

```{r}
## Manhattan plot
plot(mlm$y, cex = .4, col = "grey",
     ylab = "-log10(pvalues)",
     main = "Manhattan plot (mlm)" )
points(sim$causal, mlm$y[sim$causal], 
       type = "h", lty = 1, col = "green4")
abline(-log10(0.05/length(pv.oracle)), 0, lty = 2)
```

```{r}
qqplot(-log10(pv.oracle), mlm$y, col = "grey", cex = .4, ylab ="mlm scores")
abline(0,1, col = 'green4')
```




Let us compare the results of method that estimate confounders at the same time as it computes association with phenotype. The **cate** method implements latent factor mixed models. Based on a principal component analysis of the genotype data, we evaluate that around $K = 6$ factors explain population structure, and 6 hidden factors are used in cate. In cate, the pvalues are computed as follows. 


```{r cate, dependson=c("oracle")}   
  library(cate)
  pv.cate <- cate( ~ sim.phenotype, 
                   X.data = data.frame(sim$phenotype), 
                   Y = genotype, 
                   r = 6, 
                   calibrate = TRUE)$beta.p.value
```

Now let us visualize the Manhattan plot for the cate method and compare it to the Manhattan plot for the oracle method. The vertical bars correspond to the causal variants in the simulation. 

```{r, dependson=c("cate")}   
   plot(-log10(pv.cate), cex = .4, col = "grey", main = "Manhattan plot (Latent factor model)" )
   points(sim$causal, -log10(pv.cate)[sim$causal], type = "h", lty = 1, col = "orange" )
    abline(-log10(0.05/length(pv.oracle)), 0, lty = 2)
```

We eventually compare the results of the latent factor model with the oracle method.

```{r, dependson=c("cate")}   
   ## qqplot
   qqplot(-log10(pv.oracle), -log10(pv.cate) , pch = 19, cex = .4, col = "grey" )
   abline( 0, 1, lwd = 2, col = "orange" )
```

In this situation, the latent factor model performed similarly to the oracle method. Other testing methods could be evaluated similarly (glm, mixed models, etc).  



##### Comparison with LFMM

For comparisons with **cate**, we also use $K = 6$ factors in **lfmm**.  


```{r lfmm}   
lfmm.mod <- lfmm::lfmm_ridge(Y = genotype, X = sim$phenotype, K = 6)
```

We compute the pvalues as follows

```{r}   
  p <- lfmm::lfmm_test(Y = genotype, X = sim$phenotype, lfmm = lfmm.mod, calibrate = "gif")
  pv.lfmm <- p$calibrated.pvalue
```

Now let us visualize the Manhattan plots for **lfmm** and **cate**. The vertical bars correspond to the causal variants in the simulation. 

```{r, dependson=c("cate")}   
   plot( -log10(pv.lfmm), cex = .4, col = "grey", main = "Manhattan plot (lfmm)" )
   points( sim$causal, -log10(pv.lfmm)[sim$causal], type = "h", lty = 1, col = "blue" )
   abline(-log10(0.05/length(pv.oracle)), 0, lty = 2)
```

We eventually compare the results of the latent factor model with the oracle method.

```{r, dependson=c("cate")}   
   ## plot
   qqplot(-log10(pv.oracle), -log10(pv.lfmm) , pch = 19, cex = .4, col = "grey" )
   abline( 0, 1, lwd = 2, col = "blue" )
```




#### Package reference

* Francois O, Caye K (2017). naturalgwas: An R package for simulating phenotypes and evaluating GWAS methods with large geographic sampling.


#### References


* Atwell S, Huang YS, Vilhjalmsson BJ, et al. (2010). Genome-wide association study of 107 phenotypes in Arabidopsis thaliana inbred lines. Nature 465, 627-631.

* Francois O, Martins H, Caye K, Schoville SD (2016). Controlling false discoveries in genome scans for selection. Molecular Ecology 25, 454-469.

* Frichot E, Francois O (2015). LEA: an R package for Landscape and Ecological Association studies. Methods in Ecology and Evolution 6(8), 925-929.




